jupyterhub/tests/test_roles.py

```diff
@@ -1023,7 +1020,119 @@ async def test_config_role_users():
+    hub.init_db()
+    hub.authenticator.admin_users = ['admin']
+    hub.authenticator.allowed_users = ['eddy']
+    await hub.init_role_creation()
+    await hub.init_users()
+    await hub.init_role_assignment()
+    admin_role = orm.Role.find(hub.db, name='admin')
+    for user_name in ['eddy', 'admin']:
+        user = orm.User.find(hub.db, name=user_name)
+        assert user.admin
+        assert admin_role in user.roles
+    admin_role_spec[0]['users'].remove('eddy')
+    hub.load_roles = admin_role_spec
+    await hub.init_users()
+    await hub.init_role_assignment()
+    user = orm.User.find(hub.db, name='eddy')
+    assert not user.admin
+    assert admin_role not in user.roles
+
+
+async def test_custom_role_reset():
+    user_role_spec = [
+        {
+            'name': 'user',
+            'scopes': ['self', 'shutdown'],
+            'users': ['eddy'],
+        }
+    ]
+    hub = MockHub(load_roles=user_role_spec)
+    hub.init_db()
+    hub.authenticator.allowed_users = ['eddy']
+    await hub.init_role_creation()
+    await hub.init_users()
+    await hub.init_role_assignment()
+    user_role = orm.Role.find(hub.db, name='user')
+    user = orm.User.find(hub.db, name='eddy')
+    assert user_role in user.roles
+    assert 'shutdown' in user_role.scopes
+    hub.load_roles = []
+    await hub.init_role_creation()
+    await hub.init_users()
+    await hub.init_role_assignment()
+    user_role = orm.Role.find(hub.db, name='user')
+    user = orm.User.find(hub.db, name='eddy')
+    assert user_role in user.roles
+    assert 'shutdown' not in user_role.scopes
+
+
+async def test_removal_config_to_db():
+    role_spec = [
+        {
+            'name': 'user',
+            'scopes': ['self', 'shutdown'],
+        },
+        {
+            'name': 'wizard',
+            'scopes': ['self', 'read:groups'],
+        },
+    ]
+    hub = MockHub(load_roles=role_spec)
+    hub.init_db()
+    await hub.init_role_creation()
+    assert orm.Role.find(hub.db, 'user')
+    assert orm.Role.find(hub.db, 'wizard')
+    hub.load_roles = []
+    await hub.init_role_creation()
+    assert orm.Role.find(hub.db, 'user')
+    assert not orm.Role.find(hub.db, 'wizard')
+
+
+async def test_user_config_respects_memberships():
+    role_spec = [
+        {
+            'name': 'user',
+            'scopes': ['self', 'shutdown'],
+        }
+    ]
+    user_names = ['eddy', 'carol']
+    hub = MockHub(load_roles=role_spec)
+    hub.init_db()
+    hub.authenticator.allowed_users = user_names
+    await hub.init_role_creation()
+    await hub.init_users()
+    await hub.init_role_assignment()
+    user_role = orm.Role.find(hub.db, 'user')
+    for user_name in user_names:
+        user = orm.User.find(hub.db, user_name)
+        assert user in user_role.users
+
+
+async def test_admin_role_respects_config():
+    role_spec = [
+        {
+            'name': 'admin',
+            'scopes': ['self', 'shutdown'],
```

Comment:
Does this rewrite `admin` role's scopes? It shouldn't, the 'admin' role's definition and scopes should not be rewritable, the `create_role()` in `roles.py` should call `_overwrite_role()` which should raise error. Do the scopes get overwritten in the default `admin` role?

Reply:
Good catch, thanks! Fixed and added a test for it